src/net.h

```diff
@@ -644,6 +653,7 @@ class CNodeStats
     // Bind address of our side of the connection
     CAddress addrBind;
     uint32_t m_mapped_as;
+    std::string m_connection_type;
```

Comment:
do I understand your suggestion correctly: update `CNodeStats.m_connection_type` to the raw `ConnectionType`, then in `getpeerinfo` implementation, pass through a function `ConnectionTypeToString(m_connection_type)`? 

it's possible, what's the advantage? 

the main downside I'm seeing is I'm trying to keep `m_connection_type` private, and need to access it from `net_processing` for the log commit. so having `ConnectionTypeAsString` just return the `std::string` is more convenient.

open to suggestions

Reply:
resolving this convo for now, feel free to reopen/comment if you'd like me reconsider :)

Reply:
I think string is fine, but I think the two variables are now a bit confusing. `m_conn_type` in `CNode` is a `ConnectionType`, and `m_connection_type` in `CNodeStats` is a `string`.

My suggestion is to either

1. Switch this to a `ConnectionType`, and maybe even use the same name; this unifies the two and no one has to ever wonder what the difference is between the `m_conn_type` and the `m_connection_type`.
2. Rename this e.g. to `m_conn_string`.

Feel free to ignore, though.

Reply:
I would agree that storing a string here is a bit strange. This is a persistent structure in memory, and every string causes an extra heap allocation. Which means an extra heap allocation per node in memory. While it stores only one from a limited number of choices, so as I understand, it could just as well be an enum?

Reply:
I'm not sure this is a problem:

- in bitcoind, CNodeStats is just instantiated when needed, rather than kept persistently (it looks like the objects are cached in bitcoin-qt though)
- most compilers will use the short string optimization, and most of these strings are < 15 chars long, so won't result in a heap allocation
- there are already 3 other `std::string` members in this struct

Reply:
- definitely agree that `m_conn_string` is a better variable name for the current state, have incorporated into the latest push 
- RE memory management- I'm not fully understanding how this is a persistent data structure. I tried to trace the call sites and I'm not seeing where its got anything other than automatic storage duration. But also don't understand the gui caching. any pointers??
- regardless, I've taken a shot at implementing having `CNodeStats` store a `ConnectionType` instead of a string. I haven't incorporated into this patch yet, but happy to if this solution looks reasonable / preferable: https://github.com/amitiuttarwar/bitcoin/commit/93a56e2880ee2a1513447c1c8807ebfa70be9270. One thing is @laanwj you expressed [here](https://github.com/bitcoin/bitcoin/pull/19725#discussion_r473018587) a preference for moving the function out of the header file, which this implementation undoes. 

I don't have any strong preferences, happy with whatever reviewers prefer.